fastserde/avro-fastserde/src/main/java/org/apache/avro/generic/CustomizedGenericDatumReader.java

```diff
@@ -0,0 +1,44 @@
+package org.apache.avro.generic;
+
+import com.linkedin.avro.fastserde.customized.DatumReaderCustomization;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.ColdDatumReaderMixIn;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericDatumReader;
+
+/**
+ * This class will apply {@link DatumReaderCustomization} at runtime.
+ */
+public class CustomizedGenericDatumReader<T> extends GenericDatumReader<T> implements ColdDatumReaderMixIn {
+  private final DatumReaderCustomization customization;
+
+  public CustomizedGenericDatumReader(Schema writerSchema, Schema readerSchema, GenericData modelData, DatumReaderCustomization customization) {
+    super(writerSchema, readerSchema, modelData != null ? modelData : GenericData.get());
+    if (customization == null) {
+      throw new IllegalArgumentException("'customization' param should not null when constructing " +  this.getClass().getName());
+    }
+    this.customization = customization;
+  }
+
+  public CustomizedGenericDatumReader(Schema writerSchema, Schema readerSchema, DatumReaderCustomization customization) {
+    super(writerSchema, readerSchema);
+    if (customization == null) {
+      throw new IllegalArgumentException("'customization' param should not null when constructing " +  this.getClass().getName());
+    }
+    this.customization = customization;
+  }
+
+
+  @Override
+  protected Object newMap(Object old, int size) {
+    if (customization.getNewMapOverrideFunc() != null) {
+      return customization.getNewMapOverrideFunc().apply(old, size);
+    }
+    return super.newMap(old, size);
```

Comment:
We can make this the same as the generated code, right? Just always call:

```
return customization.getNewMapOverrideFunc().apply(old, size);
```

Reply:
Yes, we can now.
In the future, when we support more customization functions, we need to backfill all the missing functions to make sure each function won't be null, otherwise, this logic can break.

Reply:
Yes, that is how I hope it will be... the setter on the customization object should reject null closures, and by default should be populated with the vanilla implementation.